### PR TITLE
Release into master

### PR DESCRIPTION
**Pull request by PatrickDekker98**
I'm feeling bold, here's a pr

### CODE REVIEW

File: code/esp_code/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# ESP CODE
+## What esp is used?
+We have used this code with the heltec wifi kit 32. This is because one of the engineers had on lying around.
```
Comment (florianhumblot) [inline code/esp_code/README.md]: `on` should be `one`

File: code/esp_code/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# ESP CODE
+## What esp is used?
+We have used this code with the heltec wifi kit 32. This is because one of the engineers had on lying around. 
+## How to compile and flash
+This code is written with the use of the official esp_idf from espressif. 
+Here's how to get started: https://docs.espressif.com/projects/esp-idf/en/latest/get-started/ 
+Here's the github: https://github.com/espressif/esp-idf 
+If all this is done copy the tcp_client example directory next to the esp_ide directory. Then copy this tcp_client.c in place of the existing one. Next set your port with `man menuconfig`. You should now be able to flash your esp with `make flash`. Monitor it with `make monitor` close your monitor with `Crtl+]`.
```
Comment (florianhumblot) [inline code/esp_code/README.md]: There doesn't seem to be an `esp_ide` directory in this repo.

**Review by florianhumblot — APPROVED** [review]
minor nitpicks, would be nice if it was changed but I won't block this merge for it.

File: code/esp_code/README.md (hunk 1 above)
Comment (PatrickDekker98) [reply code/esp_code/README.md]: is fixed

File: code/esp_code/tcp_client.c [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+static void tcp_client_task(void *pvParameters) {
+    // char rx_buffer[128];
+    char addr_str[128];
+    int addr_family;
+    int ip_protocol;
+
+    while (1) {
+
+#ifdef CONFIG_EXAMPLE_IPV4
```
Comment (LRstudentHU) [inline code/esp_code/tcp_client.c]: We're not going to use ipv6, so you can remove these blocks from the code.

File: code/esp_code/tcp_client.c [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        assert(ret == ESP_OK);
+
+        memset(recvbuf, 0, 33);
+        spi_slave_transaction_t t;
+        memset(&t, 0, sizeof(t));
+
+        while (1) {
+            // Clear receive buffer,
+            memset(recvbuf, 0xA5, 129);
+            // sprintf(sendbuf, "This is the receiver, sending data for
```
Comment (gerritvanos) [inline code/esp_code/tcp_client.c]: is this commented code, or just clarification of the line above

File: code/esp_code/tcp_client.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        gpio_config(&io_conf);
+        // Enable pull-ups on SPI lines so we don't detect rogue pulses when no
+        // master is connected.
+        gpio_set_pull_mode(GPIO_MOSI, GPIO_PULLUP_ONLY);
+        gpio_set_pull_mode(GPIO_SCLK, GPIO_PULLUP_ONLY);
+        gpio_set_pull_mode(GPIO_CS, GPIO_PULLUP_ONLY);
+
+        // Initialize SPI slave interface
+        ret = spi_slave_initialize(HSPI_HOST, &buscfg, &slvcfg, 1);
+        assert(ret == ESP_OK);
```
Comment (LRstudentHU) [inline code/esp_code/tcp_client.c]: Other functions that can fail (e.g. `connect`) check the error with an if-statement and log on failure. Maybe you could replace this assert with that? I believe asserts aren't compiled on a production build (e.g. no debug info).

File: code/esp_code/tcp_client.c [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+        memset(recvbuf, 0, 33);
```
Comment (LRstudentHU) [inline code/esp_code/tcp_client.c]: What does the `33` represent here? `recvbuf` is a char array that is 129 bytes.

**Review by gerritvanos — COMMENTED** [review]
same as florian only small question

File: code/esp_code/tcp_client.c [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        spi_slave_transaction_t t;
```
Comment (LRstudentHU) [inline code/esp_code/tcp_client.c]: You can replace the `memset` with `spi_slave_transaction_t t = { 0 }`. Note: this is different from C++!

File: code/esp_code/tcp_client.c [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        memset(&t, 0, sizeof(t));
+
+        while (1) {
+            // Clear receive buffer,
+            memset(recvbuf, 0xA5, 129);
```
Comment (LRstudentHU) [inline code/esp_code/tcp_client.c]: Why `0xA5`?

File: code/headers/frame.hpp [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+     */
+    struct frame_s {
+        /**
+         * @brief Size of the data in bytes (1 byte, 8 bits equals 1)
+         */
+        size_t size;
+        /**
+         * @brief Frame data including headers
+         */
+        uint8_t data[];
```
Comment (LRstudentHU) [inline code/headers/frame.hpp:18]: Maybe a `std::array` would be a bit neater?

File: code/src/esp_32.cpp [hunk 10]
```diff
@@ -0,0 +1,29 @@
+#include <esp_32.hpp>
+
+namespace r2d2::communication {
+
+    esp_32_c::esp_32_c(hwlib::spi_bus &spi_connection, hwlib::pin_out &ss)
+        : spi_connection(spi_connection), slave_select(ss) {
+    }
+
+    void esp_32_c::send(const r2d2::frame_external_s &frame) {
+        uint8_t recv[256] = {0};
```
Comment (LRstudentHU) [inline code/src/esp_32.cpp:12]: In C++, `= { 0 }` on an array will only set the first element of the array to zero. You will want `= {}`.

File: test/main.cpp [hunk 11]
```diff
@@ -0,0 +1,27 @@
+#include "ostream"
+
+#define CATCH_CONFIG_MAIN
+#include <catch.hpp>
+#include <mock_external.hpp>
+
+TEST_CASE("mock_external_c, set_send_reaction compiles", ""){
+    auto test_mock = r2d2::communication::mock_external_c();
+    auto frame = r2d2::frame_external_s();
+    std::function<void(
```
Comment (LRstudentHU) [inline test/main.cpp:10]: The formatting makes this really hard to read. I know it is probably clang format that is causing this, but could you manually format this piece of code?

File: test/main.cpp [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+            const r2d2::frame_external_s&, 
+            std::vector<r2d2::frame_external_s>&
+            )> lam = [&](
+                const r2d2::frame_external_s &fr, 
+                std::vector<r2d2::frame_external_s> &retvec
+                )->void{
```
Comment (LRstudentHU) [inline test/main.cpp:16]: `-> void` as a lambda return type is inferred and can be omitted.

**Review by LRstudentHU — CHANGES_REQUESTED** [review]
Could you please replace the gitignore with the one from the template and add the .ipch files to it? Currently there are large binary files in this PR that shouldn't really be there.

File: test/main.cpp (hunk 11 above)
Comment (PatrickDekker98) [reply test/main.cpp:10]: I am aware, however I believe this is the most readable way this code can be written

File: test/main.cpp (hunk 12 above)
Comment (PatrickDekker98) [reply test/main.cpp:16]: However it does not have to be

File: code/src/esp_32.cpp (hunk 10 above)
Comment (PatrickDekker98) [reply code/src/esp_32.cpp:12]: I only  need the fist element of the array to be zero, this is the length of the external frame, if this is zero I can assume I received nothing

File: code/esp_code/tcp_client.c (hunk 3 above)
Comment (PatrickDekker98) [reply code/esp_code/tcp_client.c]: I will, but do you mind if I create an issue and fix this next iteration?

Comment (LRstudentHU) [reply code/esp_code/tcp_client.c]: I don't think this example/dead code should be in master

File: code/src/esp_32.cpp (hunk 10 above)
Comment (LRstudentHU) [reply code/src/esp_32.cpp:12]: Could you clarify that in a comment?

File: code/esp_code/README.md (hunk 2 above)
Comment (PatrickDekker98) [reply code/esp_code/README.md]: fixed this

File: code/esp_code/tcp_client.c (hunk 4 above)
Comment (PatrickDekker98) [reply code/esp_code/tcp_client.c]: commend was removed

File: code/esp_code/tcp_client.c (hunk 3 above)
Comment (PatrickDekker98) [reply code/esp_code/tcp_client.c]: I removed it anyways in #4

File: code/src/esp_32.cpp (hunk 10 above)
Comment (PatrickDekker98) [reply code/src/esp_32.cpp:12]:
```c++
struct frame_external_s {
        uint8_t length;
        external_id_s id;
        frame_type type;

        // NOTE: data should come last; ordering is important
        // for this specific struct!
        uint8_t data[256 - sizeof(uint8_t) - sizeof(external_id_s) - sizeof(frame_type)];
};
```
See, the first byte of frame_external_s is its length, an frame without a length contains nothing or did I misunderstand? 
(https://github.com/R2D2-2019/communication/blob/54129b0274da275c574afbf04e155042d91d616a/code/src/esp_32.cpp#L15) 
Here I check whether the first byte received is zero, if zero i do not write the frame into the que

Comment (LRstudentHU) [reply code/src/esp_32.cpp:12]: I meant in the code, to clarify that you only want the first byte :-P

Comment (PatrickDekker98) [reply code/src/esp_32.cpp:12]: aha

**Review by LRstudentHU — APPROVED** [review]

**Pull request merged**